### ISSUE TITLE
Hashtable Lookup Panic

### ISSUE THREAD

**Issue by ethanpailes** [lalrpop/lalrpop#391]
I ran into a panic in the lalrpop executable.

# Steps to Reproduce

1. `git clone git@github.com:lalrpop/lalrpop.git`
2. `cd lalrpop/lalrpop && build`
3. Create a file called `lalrpop_panic.lalrpop` which contains
```
grammar;

pub BlockBody: ast::Expr = ExprNode<BlockBodyKind>;

//////////////////////////////////////////////////////////////////////////
//                                                                      //
//                            Expressions                               //
//                                                                      //
//////////////////////////////////////////////////////////////////////////

Expr0: ast::ExprKind = {
    "{" <ss: StatementNode<Statement0>> <e: ExprNode<Expr0>> "}" =>
        ast::ExprKind::Block(s, Box::new(e)),

    Expr1,
};

Expr1: ast::ExprKind = {
    <l: ExprNode<Expr1>> "&&" <r: ExprNode<Expr5>> =>
        ast::ExprKind::BinOp(Box::new(l), ast::BOp::And, Box::new(r)),

    Expr5,
};

Expr5: ast::ExprKind = {
    // if expressions
    "if" "(" <cond: ExprNode<Expr0>> ")" <tl: @L> "{"
        <true_branch: BlockBody> <tr: @R>
    "}" "else" <fl: @L> "{"
        <false_branch: BlockBody>
    "}" <fr: @R> => undefined!(),

    "(" <e: Expr0> ")" => e,
};

BlockBodyKind: ast::ExprKind = {
    <s: (StatementNode<Statement0>)*> <e: ExprNode<Expr1>> =>
        ast::ExprKind::Block(s, Box::new(e)),
};

//////////////////////////////////////////////////////////////////////////
//                                                                      //
//                            Statements                                //
//                                                                      //
//////////////////////////////////////////////////////////////////////////

Statement0: ast::StatementKind = {
    <e: ExprNode<Expr0>> ";" => ast::StatementKind::Expr(Box::new(e)),
};

//////////////////////////////////////////////////////////////////////////
//                                                                      //
//                            Macros                                    //
//                                                                      //
//////////////////////////////////////////////////////////////////////////

ExprNode<KindRule>: ast::Expr = {
    <l: @L> <k: KindRule> <r: @R> =>
        ast::Expr::new(k, ast::Span { start: l, end: r }),
};

StatementNode<KindRule>: ast::Statement = {
    <l: @L> <k: KindRule> <r: @R> =>
        ast::Statement::new(k, ast::Span { start: l, end: r }),
};
```
4. `cargo run lalrpop_panic.lalrpop`

# Expected Behavior

I would expect lalrpop to either output a parser rust file, or print an error message explaining how the above grammar is wrong.

# Actual behavior

It prints
```
    Finished dev [unoptimized + debuginfo] target(s) in 0.07s
     Running `/tmp/lalrpop/target/debug/lalrpop ../lalrpop_panic.lalrpop`
processing file `../lalrpop_panic.lalrpop`
thread 'main' panicked at 'no entry found for key', libcore/option.rs:914:5
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

# Additional Notes

When I asked for a backtrace, it told me the problem was coming from [here](https://github.com/lalrpop/lalrpop/blob/b87e606372e953353a29a6c0fe921138d64b2973/lalrpop/src/lr1/lane_table/construct/merge.rs#L205).

I added a few print statements to dump the state of the hashmap and the key right before that line is executed, which had the following result:
```
    Finished dev [unoptimized + debuginfo] target(s) in 0.06s
     Running `/tmp/lalrpop/target/debug/lalrpop ../lalrpop_panic.lalrpop`
processing file `../lalrpop_panic.lalrpop`
state_sets={S0: StateSet { index: 0 }, S1: StateSet { index: 1 }, S12: StateSet { index: 2 }, S14: StateSet { index: 3 }, S21: StateSet { index: 4 }, S22: StateSet { index: 5 }, S30: StateSet { index: 6 }, S34: StateSet { index: 7 }}
target=S1
state_sets={S0: StateSet { index: 0 }, S1: StateSet { index: 1 }, S12: StateSet { index: 2 }, S14: StateSet { index: 3 }, S21: StateSet { index: 4 }, S22: StateSet { index: 5 }, S30: StateSet { index: 6 }, S34: StateSet { index: 7 }}
target=S5
thread 'main' panicked at 'no entry found for key', libcore/option.rs:914:5
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

I'm afraid I don't know enough about the code to make much sense of this, but maybe it will be useful to someone with more knowledge.

**Comment by Marwes** [comment]
Another error case

```


grammar;

pub Expr: Box<Expr> = {
    Bind,
    "func" "{" (Arm ",")* "}",
    Term+,
};

Bind: Bind = {
    <Pattern> "=" <Expr> ";",
};

Term: Box<Expr> = {
    Primitive,
    "(" Expr ")",
};

Arm: (Pattern, Expr) = {
    Pattern "=>" Expr,
};

Pattern = {
    Primitive,
    "(" (Pattern ",")+ Pattern ")",
};

Primitive = {
    r"-?[0-9]*",
};

```

**Comment by Storyyeller** [comment]
I'm running into the same issue with a much smaller grammar:

`grammar();

pub S = {
    "a" E "c",
    "a" F "d",
    "b" F "c",
    "b" E "d",
}

E = { "x" }
F = { "x" }`

**Comment by modulovalue** [comment]
The smaller grammar by Storyyeller [on grammophone](https://mdaines.github.io/grammophone/?s=UyAtPiAiYSIgRSAiYyIKICAgfCAiYSIgRiAiZCIKICAgfCAiYiIgRiAiYyIKICAgfCAiYiIgRSAiZCIuCkUgLT4gIngiLgpGIC0+ICJ4Ii4=)

It looks pretty much like the grammar from https://github.com/lalrpop/lalrpop/issues/768#issuecomment-1787079315 so that issue and this one are possibly duplicates.

**Comment by dburgener** [comment]
This should be fixed by #898.  Thanks for the report!  If you're still seeing issues after the next release (or when building the lastest code from github) please reopen.
